README.md
# Ansible Collection - pytoccaz.azure

<!--start requires_ansible-->
## Ansible version compatibility

This collection has been tested against following Ansible versions: **>=2.13.13**.

For collections that support Ansible 2.9, please ensure you update your `network_os` to use the
fully qualified collection name (for example, `cisco.ios.ios`).
Plugins and modules within a collection may be tested with only specific Ansible versions.
A collection may contain metadata that identifies these versions.
PEP440 is the schema used to describe the versions of Ansible.
<!--end requires_ansible-->

This collection has been tested against `azure.azcollection` version **1.18.1**.

## Installation

Collection `azure.azcollection` is required. 

Download from Galaxy:

```bash
ansible-galaxy collection install pytoccaz.azure
```

## Collection content

<!--start collection content-->
### Modules
Name | Description
--- | ---
[pytoccaz.azure.az_blobcontainer_list](https://github.com/pytoccaz/ansible_azure/blob/main/docs/pytoccaz.azure.az_blobcontainer_list_module.rst)|List Azure blobs module

<!--end collection content-->

## Credits
This program is inspired by and contains code snippets from `azure.azcollection.azure_rm_storageshare` module by Andrii Bilorus.

It also contains documentation fragments from `azure.azcollection.azure` doc_fragments file by Matt Davis and Chris Houseknecht.


CHANGELOG.md
## v1.1.0 (2023-10-25)

- Add new arg datetime_format to az_blobcontainer_list module to control last_modified date format
- Change blob last_modified date default format to match to_datetime ansible filter format.
- Fix some sanity checks

## v1.0.0 (2023-10-23)

- Major release for publishing

## v0.0.1 (2023-10-23)

- Init release

plugins/doc_fragments/azure.py
# -*- coding: utf-8 -*-

# Copyright: (c) 2023 Olivier Bernard
# Copyright: (c) 2016 Matt Davis, <mdavis@ansible.com>
# Copyright: (c) 2016 Chris Houseknecht, <house@redhat.com>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function
__metaclass__ = type


class ModuleDocFragment(object):

    # Azure doc fragment
    DOCUMENTATION = r'''

options:
    ad_user:
        description:
            - Active Directory username. Use when authenticating with an Active Directory user rather than service
              principal.
        type: str
    password:
        description:
            - Active Directory user password. Use when authenticating with an Active Directory user rather than service
              principal.
        type: str
    profile:
        description:
            - Security profile found in ~/.azure/credentials file.
        type: str
    subscription_id:
        description:
            - Your Azure subscription Id.
        type: str
    client_id:
        description:
            - Azure client ID. Use when authenticating with a Service Principal.
        type: str
    secret:
        description:
            - Azure client secret. Use when authenticating with a Service Principal.
        type: str
    tenant:
        description:
            - Azure tenant ID. Use when authenticating with a Service Principal.
        type: str
    cloud_environment:
        description:
            - For cloud environments other than the US public cloud, the environment name (as defined by Azure Python SDK, eg, C(AzureChinaCloud),
              C(AzureUSGovernment)), or a metadata discovery endpoint URL (required for Azure Stack). Can also be set via credential file profile or
              the C(AZURE_CLOUD_ENVIRONMENT) environment variable.
        type: str
        default: AzureCloud
    adfs_authority_url:
        description:
            - Azure AD authority url. Use when authenticating with Username/password, and has your own ADFS authority.
        type: str
    cert_validation_mode:
        description:
            - Controls the certificate validation behavior for Azure endpoints. By default, all modules will validate the server certificate, but
              when an HTTPS proxy is in use, or against Azure Stack, it may be necessary to disable this behavior by passing C(ignore). Can also be
              set via credential file profile or the C(AZURE_CERT_VALIDATION) environment variable.
        type: str
        choices: [ ignore, validate ]
    auth_source:
        description:
            - Controls the source of the credentials to use for authentication.
            - Can also be set via the C(ANSIBLE_AZURE_AUTH_SOURCE) environment variable.
            - When set to C(auto) (the default) the precedence is module parameters -> C(env) -> C(credential_file) -> C(cli).
            - When set to C(env), the credentials will be read from the environment variables
            - When set to C(credential_file), it will read the profile from C(~/.azure/credentials).
            - When set to C(cli), the credentials will be sources from the Azure CLI profile. C(subscription_id) or the environment variable
              C(AZURE_SUBSCRIPTION_ID) can be used to identify the subscription ID if more than one is present otherwise the default
              az cli subscription is used.
            - When set to C(msi), the host machine must be an azure resource with an enabled MSI extension. C(subscription_id) or the
              environment variable C(AZURE_SUBSCRIPTION_ID) can be used to identify the subscription ID if the resource is granted
              access to more than one subscription, otherwise the first subscription is chosen.
            - The C(msi) was added in Ansible 2.6.
        type: str
        default: auto
        choices:
        - auto
        - cli
        - credential_file
        - env
        - msi
    api_profile:
        description:
        - Selects an API profile to use when communicating with Azure services. Default value of C(latest) is appropriate for public clouds;
          future values will allow use with Azure Stack.
        type: str
        default: latest
    log_path:
        description:
            - Parent argument.
        type: str
    log_mode:
        description:
            - Parent argument.
        type: str
    x509_certificate_path:
        description:
            - Path to the X509 certificate used to create the service principal in PEM format.
            - The certificate must be appended to the private key.
            - Use when authenticating with a Service Principal.
        type: path
    thumbprint:
        description:
            - The thumbprint of the private key specified in I(x509_certificate_path).
            - Use when authenticating with a Service Principal.
            - Required if I(x509_certificate_path) is defined.
        type: str
requirements:
    - python >= 2.7
    - The host that executes this module must have the azure.azcollection collection installed via galaxy
    - All python packages listed in collection's requirements-azure.txt must be installed via pip on the host that executes modules from azure.azcollection
    - Full installation instructions may be found https://galaxy.ansible.com/azure/azcollection

notes:
    - For authentication with Azure you can pass parameters, set environment variables, use a profile stored
      in ~/.azure/credentials, or log in before you run your tasks or playbook with C(az login).
    - Authentication is also possible using a service principal or Active Directory user.
    - To authenticate via service principal, pass subscription_id, client_id, secret and tenant or set environment
      variables AZURE_SUBSCRIPTION_ID, AZURE_CLIENT_ID, AZURE_SECRET and AZURE_TENANT.
    - To authenticate via Active Directory user, pass ad_user and password, or set AZURE_AD_USER and
      AZURE_PASSWORD in the environment.
    - "Alternatively, credentials can be stored in ~/.azure/credentials. This is an ini file containing
      a [default] section and the following keys: subscription_id, client_id, secret and tenant or
      subscription_id, ad_user and password. It is also possible to add additional profiles. Specify the profile
      by passing profile or setting AZURE_PROFILE in the environment."

seealso:
    - name: Sign in with Azure CLI
      link: https://docs.microsoft.com/en-us/cli/azure/authenticate-azure-cli?view=azure-cli-latest
      description: How to authenticate using the C(az login) command.
    '''


plugins/modules/az_blobcontainer_list.py
#!/usr/bin/python

# Copyright: (c) 2023, Olivier BERNARD
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
from __future__ import (absolute_import, division, print_function)

__metaclass__ = type

DOCUMENTATION = r'''
---
module: az_blobcontainer_list

short_description: List Azure blobs module

version_added: "1.0.0"

description: Lists the blobs under a specified azure storage container.

options:
    resource_group:
        description:
            - Name of the resource group to use.
        required: true
        aliases:
            - resource_group_name
        version_added: '1.0.0'
    storage_account_name:
        description:
            - Name of the storage account to use.
        required: true
        aliases:
            - account_name
            - storage_account
        version_added: '1.0.0'
    container:
        description:
            - Name of a blob container within the storage account.
        required: true
        aliases:
            - container_name
        version_added: '1.0.0'
    name_starts_with:
        description:
            - Filters the results to return only blobs whose names begin with the specified prefix.
        required: false
        aliases:
            - start_with
        version_added: '1.0.0'
    datetime_format:
        description:
            - Rendering format for last_modified container and blobs date.
        required: false
        default: '%Y-%m-%d %H:%M:%S'
        aliases:
            - dt_format
        version_added: '1.1.0'

extends_documentation_fragment:
    - pytoccaz.azure.azure

author:
    - Olivier Bernard (@pytoccaz)

'''

EXAMPLES = r'''
# List blobs under a container
- name: List blobs under mycontainer
  az_blobcontainer_list:
    resource_group: myresourcegroup
    account_name: myaccountname
    container: mycontainer

'''

RETURN = r'''
blobs:
    description:
        - List of blobs.
    returned: always
    type: list
    sample:
        [
            {
                "content_length": 136532,
                "content_settings": {
                    "cache_control": null,
                    "content_disposition": null,
                    "content_encoding": null,
                    "content_language": null,
                    "content_md5": null,
                    "content_type": "application/image"
                },
                "last_modified": "09-Mar-2016 22:08:25",
                "name": "foo.png",
                "tags": {},
                "type": "BlockBlob"
            }
        ]
container:
    description:
        - Facts about the current state of the selected container.
    returned: always
    type: dict
    sample: {
        "last_modified": "2016-03-09 19:28:26",
        "name": "foo",
        "tags": {}
    }

'''
from ansible_collections.azure.azcollection.plugins.module_utils.azure_rm_common import AzureRMModuleBase
import re

try:
    from azure.core.exceptions import ResourceNotFoundError
except ImportError:
    # This is handled in azure_rm_common
    pass


class AzureBlobContainerList(AzureRMModuleBase):

    def __init__(self):

        self.module_arg_spec = dict(
            storage_account_name=dict(required=True, type='str', aliases=[
                                      'account_name', 'storage_account']),
            container=dict(required=True, type='str',
                           aliases=['container_name']),
            resource_group=dict(required=True, type='str',
                                aliases=['resource_group_name']),
            name_starts_with=dict(required=False, type='str',
                                  aliases=['starts_with']),
            datetime_format=dict(required=False, type='str',
                                 default='%Y-%m-%d %H:%M:%S',
                                 aliases=['dt_format']),
        )
        self.resource_group = None
        self.storage_account_name = None
        self.container = None
        self.container_client = None
        self.name_starts_with = None
        self.results = dict(
            changed=False,
            container=dict(),
            blobs=list(),
        )
        self.valide_container_name = re.compile(
            "^[a-z0-9]([a-z0-9]|-(?!-)){1,61}[a-z0-9]$")
        self.datetime_format = None

        super(AzureBlobContainerList, self).__init__(derived_arg_spec=self.module_arg_spec,
                                                     supports_check_mode=True,
                                                     supports_tags=False)

    def exec_module(self, **kwargs):

        for key in list(self.module_arg_spec.keys()):
            setattr(self, key, kwargs[key])

        if not self.valide_container_name.match(self.container):
            self.fail(
                "Container name {0} is not a valid string".format(
                    self.container)
            )

        self.results['check_mode'] = self.check_mode

        self.results["container"] = self.container

        self.blob_service_client = self.get_blob_service_client(
            self.resource_group, self.storage_account_name)

        self.container_client = self.blob_service_client.get_container_client(
            container=self.container)

        try:
            container_props = self.container_client.get_container_properties()

        except ResourceNotFoundError:
            self.fail("Container {0} not found".format(self.container))

        self.results["container"] = dict(
            name=container_props["name"],
            tags=container_props["metadata"],
            last_modified=container_props["last_modified"].strftime(
                self.datetime_format),
        )

        blob_list = self.container_client.list_blobs(
            name_starts_with=self.name_starts_with, include="metadata")
        for blob in blob_list:
            blob_result = dict(
                name=blob["name"],
                tags=blob["metadata"],
                last_modified=blob["last_modified"].strftime(
                    self.datetime_format),
                type=blob["blob_type"],
                content_length=blob["size"],
                content_settings=dict(
                    content_type=blob["content_settings"]["content_type"],
                    content_encoding=blob["content_settings"]["content_encoding"],
                    content_language=blob["content_settings"]["content_language"],
                    content_disposition=blob["content_settings"]["content_disposition"],
                    cache_control=blob["content_settings"]["cache_control"],
                    content_md5=blob["content_settings"]["content_md5"].hex(
                    ) if blob["content_settings"]["content_md5"] else None,
                )
            )

            self.results["blobs"].append(blob_result)

        return self.results


def main():
    AzureBlobContainerList()


if __name__ == '__main__':
    main()
